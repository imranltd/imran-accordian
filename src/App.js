import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import fetchData from './Api';
import Accordian from './components/Accordian';
import Spinner from './components/Spinner';

const API_URL = 'https://api.myjson.com/bins/jw3rg';

const App = ({ data }) => {
  const initialLoading = (!Object.keys(data).length);

  const [accordianData, setData] = useState(data);
  const [loading, setLoading] = useState(initialLoading);

  const makeApiCall = async (url) => {
    const response = await fetchData(url);
    setData(response);
    setLoading(false);
  };

  useEffect(() => {
    if (loading) {
      makeApiCall(API_URL);
    }
  }, [loading, data]);


  const [showAnswer, setShowAnswer] = useState(null);

  const clickHandle = (e, currentState) => ((e, currentState)
    ? setShowAnswer(null)
    : setShowAnswer(e.target.id));

  return (!loading && data)
    ? (
      <>
        { accordianData.faqs.map(faq => (
          <Accordian
            key={`id-${faq.id}`}
            id={`id-${faq.id}`}
            heading={faq.question}
            details={faq.answer}
            data={faq}
            clickHandle={clickHandle}
            isOpen={(showAnswer === `id-${faq.id}`)}
            prefix="Q: "
          />
        ))
      }
      </>
    )
    : (<Spinner />);
};

App.propTypes = {
  data: PropTypes.shape({}),
};

App.defaultProps = {
  data: {},
};

export default App;
