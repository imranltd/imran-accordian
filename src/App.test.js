import React from 'react';
import renderer from 'react-test-renderer';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';

import App from './App';

describe('<App />', () => {
  const data = {
    faqs: [{
      id: 1,
      answer: 'This is an Answer',
      question: 'This is a question',
    }],
  };

  it('renders correctly', () => {
    const tree = renderer
      .create(<App data={data} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with data', () => {
    const wrapper = mount(<App data={data} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot with no data', () => {
    const wrapper = mount(<App />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
