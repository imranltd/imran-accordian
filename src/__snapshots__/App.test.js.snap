// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
.c0 {
  border-radius: 5px;
  font-family: Arial,Helvetica,sans-serif;
  margin: 10px 0;
  padding: 10px;
}

.c1 {
  -webkit-align-items: right;
  -webkit-box-align: right;
  -ms-flex-align: right;
  align-items: right;
  background-color: #23D2DA;
  border-radius: 5px;
  border: 1px solid black;
  color: #1375E6;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 10px;
}

.c1:after {
  content: "[+]";
}

.c2 {
  border-color: black;
  border-radius: 0 0 5px 5px;
  border-style: solid;
  border-width: 0 1px 1px 1px;
  display: none;
  padding: 20px 10px;
}

<div
  className="c0"
>
  <div
    className="headingClass c1"
    data-test-id="accordion-0-question"
    id="id-1"
    onClick={[Function]}
  >
    Q: 
    This is a question
  </div>
  <div
    className="detailClass c2"
    data-test-id="accordion-0-answer"
  >
    This is an Answer
  </div>
</div>
`;

exports[`<App /> should match snapshot with data 1`] = `
.c0 {
  border-radius: 5px;
  font-family: Arial,Helvetica,sans-serif;
  margin: 10px 0;
  padding: 10px;
}

.c1 {
  -webkit-align-items: right;
  -webkit-box-align: right;
  -ms-flex-align: right;
  align-items: right;
  background-color: #23D2DA;
  border-radius: 5px;
  border: 1px solid black;
  color: #1375E6;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 10px;
}

.c1:after {
  content: "[+]";
}

.c2 {
  border-color: black;
  border-radius: 0 0 5px 5px;
  border-style: solid;
  border-width: 0 1px 1px 1px;
  display: none;
  padding: 20px 10px;
}

<App
  data={
    Object {
      "faqs": Array [
        Object {
          "answer": "This is an Answer",
          "id": 1,
          "question": "This is a question",
        },
      ],
    }
  }
>
  <Accordian
    clickHandle={[Function]}
    data={
      Object {
        "answer": "This is an Answer",
        "id": 1,
        "question": "This is a question",
      }
    }
    dataId="accordion-0"
    details="This is an Answer"
    heading="This is a question"
    id="id-1"
    isOpen={false}
    key="id-1"
    prefix="Q: "
  >
    <styled.div
      isOpen={false}
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": true,
              "lastClassName": "c0",
              "rules": Array [
                "
  border-radius: 5px;
  font-family: Arial, Helvetica, sans-serif;
  margin: 10px 0;
  padding: 10px;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        isOpen={false}
      >
        <div
          className="c0"
        >
          <styled.div
            className="headingClass"
            data-test-id="accordion-0-question"
            id="id-1"
            isOpen={false}
            onClick={[Function]}
          >
            <StyledComponent
              className="headingClass"
              data-test-id="accordion-0-question"
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bwzfXH",
                    "isStatic": false,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
  align-items: right;
  background-color: #23D2DA;
  border-radius: ",
                      [Function],
                      "
  border: 1px solid black;
  color: #1375E6;
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  padding: 10px;

  &:after {
    content: ",
                      [Function],
                      "
  }
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bwzfXH",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              id="id-1"
              isOpen={false}
              onClick={[Function]}
            >
              <div
                className="headingClass c1"
                data-test-id="accordion-0-question"
                id="id-1"
                onClick={[Function]}
              >
                Q: 
                This is a question
              </div>
            </StyledComponent>
          </styled.div>
          <styled.div
            className="detailClass"
            data-test-id="accordion-0-answer"
            isOpen={false}
          >
            <StyledComponent
              className="detailClass"
              data-test-id="accordion-0-answer"
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-htpNat",
                    "isStatic": false,
                    "lastClassName": "c2",
                    "rules": Array [
                      "
  border-color: black;
  border-radius: 0 0 5px 5px;
  border-style: solid;
  border-width: 0 1px 1px 1px;
  display: ",
                      [Function],
                      "
  padding: 20px 10px;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-htpNat",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              isOpen={false}
            >
              <div
                className="detailClass c2"
                data-test-id="accordion-0-answer"
              >
                This is an Answer
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </Accordian>
</App>
`;

exports[`<App /> should match snapshot with no data 1`] = `
.c0 {
  -webkit-animation: rotate 2s linear infinite;
  animation: rotate 2s linear infinite;
  height: 50px;
  width: 50px;
}

.c0 .path {
  -webkit-animation: dash 1.5s ease-in-out infinite;
  animation: dash 1.5s ease-in-out infinite;
  stroke-linecap: round;
  stroke: red;
}

<App
  data={Object {}}
>
  <Spinner>
    <styled.svg
      viewBox="0 0 50 50"
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": true,
              "lastClassName": "c0",
              "rules": Array [
                "
  animation: rotate 2s linear infinite;
  height: 50px;
  width: 50px;

  & .path {
    animation: dash 1.5s ease-in-out infinite;
    stroke-linecap: round;
    stroke: red;
  }

  @keyframes rotate {
    100% {
      transform: rotate(360deg);
    }
  }
  @keyframes dash {
    0% {
      stroke-dasharray: 1, 150;
      stroke-dashoffset: 0;
    }
    50% {
      stroke-dasharray: 90, 150;
      stroke-dashoffset: -35;
    }
    100% {
      stroke-dasharray: 90, 150;
      stroke-dashoffset: -124;
    }
  }
",
              ],
            },
            "displayName": "styled.svg",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "svg",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        viewBox="0 0 50 50"
      >
        <svg
          className="c0"
          viewBox="0 0 50 50"
        >
          <circle
            className="path"
            cx="25"
            cy="25"
            fill="none"
            r="20"
            strokeWidth="4"
          />
        </svg>
      </StyledComponent>
    </styled.svg>
  </Spinner>
</App>
`;
